lib/checkstl.cpp

```diff
@@ -741,6 +741,11 @@ static ValueFlow::Value getLifetimeIteratorValue(const Token* tok, MathLib::bigi
     }
     if (values.size() == 1)
         return values.front();
+    it = std::find_if(values.cbegin(), values.cend(), [](const ValueFlow::Value& v) {
```

Comment:
This is problematic if there is divergent lifetimes, that is why we check the size is 1 or no duplicates for iterator kinds. Why is there multiple lifetimes?  Do we have a lifetime for `t.s.a` and `a` both?

Reply:
> This is problematic if there is divergent lifetimes, that is why we check the size is 1 or no duplicates for iterator kinds. Why is there multiple lifetimes? Do we have a lifetime for `t.s.a` and `a` both?

In the `s.a` example, the lifetime tokvalues are `s` and `.`. But I think it makes sense to chose the dot, unless there can be multiple dots.

Reply:
Its the `.` in `s.a` I assume? Ideally we should remove the lifetimes that refer to a sub expression of another lifetime from the vector. Something like this function could be called here(its not tested though):

```cpp
std::vector<ValueFlow::Value> pruneLifetimes(std::vector<ValueFlow::Value> lifetimes)
{
    std::vector<ValueFlow::Value> result;
    auto start = lifetimes.begin();
    while(start != lifetimes.end())
    {
        const Token* tok1 = start->tokvalue;
        auto it = std::partition(start, lifetimes.end(), [&](const ValueFlow::Value& v) { 
            const Token* tok2 = v.tokvalue;
            return astHasToken(tok1, tok2) or astHasToken(tok2, tok1); 
        });
        auto root = std::min_element(start, it, [](const ValueFlow::Value& x, const ValueFlow::Value& y) {
            return astHasToken(x.tokvalue, y.tokvalue);
        });
        result.push_back(*root);
        start = it;
    }
    return result;
}
```

Reply:
Not entirely sure what it does, but it seems to work :+1:

Reply:
First it groups the lifetimes together using `std::partition` with the lifetimes that refer to the same token or token of a subexpression. Then it finds the lifetime that refers to the "highest" parent using `std::min_element` and adds that to the vector. Hopefully that makes sense.